all: stop using pErr above the senders layer

andreimatei opened the pull request:
Tobias, this time I went ground-up. Changed the interface of `Txn` to return `error` everywhere, and from there on and on and on. Notably `DB.Txn()` now returns `error` too, and its closure returns `error`.
I didn't go all the way with the refactoring, there's still some leaves where I convert the `error` back to `pErr`. If it looks good, I'll go all the way, although I'm not sure how I'll ever get this merged. Might have to freeze the repo :).
Didn't run any tests yet. Started converting the KV tests over, but it takes time.

There was a place where I needed to convert a `pErr` to `error` and then back - in `Store.Send()`, which still needs to return `pErr`, it sometimes runs other batches, and takes an error from there, and then passes it back as the error for the original txn. So I've made `roachpb.NewError` recognize retryable errors and produce a `pErr` that's also retryable -> so that it will again be converted to a retryable `error`. Please see if this looks correct to you.

Otherwise the change was mostly mechanical. `pErr` has an index for its position within the batch, so I had to replace that with some code where it was used (see calls to `convertBatchError`).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6367)

<!-- Reviewable:end -->

andreimatei committed:
rename Result.PErr to Result.Err

andreimatei committed:
all: stop using pErr above the senders layer

The problem with using pErr is that any time you want to add a typed
error, you have to do it as an ErrorDetail in the proto. It's cumbersome
and you hanve no control over the scope. There was no good reason for
using pErr above the Senders except that the DB.Txn used to return pErr.

This change the interface of Txn functions and DB.Txn() to return error
instead of pErr, and from there on a ton all the dominos fall up the
stack. Retryable KV errors are now represented by a RetryableTxnError
type that is recognized by DB.Txn().

Still to come is taking the ErrorDetails that don't need to go over the
network out of the proto.

This change cuts the transition from pErr to error short in a few places
- notably some sql statements and the SQL Executor still deal with pErr
for the purposes of keeping this change under some size limit. This
means that a number of hacks have been put it to be able to go back
and forth between pErr and error.

vivekmenezes commented:
What are your final goals here? What do you want to get rid of?

On Wed, Apr 27, 2016 at 8:27 PM Andrei Matei <notifications@github.com>
wrote:

> Tobias, this time I went ground-up. Changed the interface of Txn to
> return error everywhere, and from there on and on and on. Notably DB.Txn()
> now returns error too, and its closure returns error.
> I didn't go all the way with the refactoring, there's still some leaves
> where I convert the error back to pErr. If it looks good, I'll go all the
> way, although I'm not sure how I'll ever get this merged. Might have to
> freeze the repo :).
> Didn't run any tests yet. Started converting the KV tests over, but it
> takes time.
>
> There was a place where I needed to convert a pErr to error and then back
> - in Store.Send(), which still needs to return pErr, it sometimes runs
> other batches, and takes an error from there, and then passes it back as
> the error for the original txn. So I've made roachpb.NewError recognize
> retryable errors and produce a pErr that's also retryable -> so that it
> will again be converted to a retryable error. Please see if this looks
> correct to you.
>
> Otherwise the change was mostly mechanical. pErr has an index for its
> position within the batch, so I had to replace that with some code where it
> was used (see calls to convertBatchError).
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/cockroachdb/cockroach/pull/6367
> Commit Summary
>
>    - rename Result.PErr to Result.Err
>    - wip
>    - everything but the tests builds
>
> File Changes
>
>    - *M* cli/kv.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-0> (10)
>    - *M* client/batch.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-1> (42)
>    - *M* client/client_test.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-2> (4)
>    - *M* client/db.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-3>
>    (156)
>    - *M* client/txn.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-4>
>    (161)
>    - *M* kv/split_test.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-5> (20)
>    - *M* kv/txn_coord_sender.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-6> (12)
>    - *M* kv/txn_coord_sender_test.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-7>
>    (155)
>    - *M* kv/txn_correctness_test.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-8> (82)
>    - *M* kv/txn_test.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-9>
>    (325)
>    - *M* roachpb/errors.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-10>
>    (42)
>    - *M* server/admin.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-11>
>    (12)
>    - *M* server/node.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-12> (5)
>    - *M* server/status.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-13>
>    (20)
>    - *M* server/testserver.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-14> (6)
>    - *M* sql/alter_table.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-15>
>    (71)
>    - *M* sql/backfill.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-16>
>    (148)
>    - *M* sql/create.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-17>
>    (70)
>    - *M* sql/database.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-18>
>    (20)
>    - *M* sql/delete.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-19>
>    (22)
>    - *M* sql/descriptor.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-20>
>    (61)
>    - *M* sql/drop.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-21>
>    (107)
>    - *M* sql/errors.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-22>
>    (39)
>    - *M* sql/event_log.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-23> (7)
>    - *M* sql/executor.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-24>
>    (63)
>    - *M* sql/grant.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-25>
>    (11)
>    - *M* sql/internal.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-26>
>    (15)
>    - *M* sql/kvfetcher.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-27>
>    (14)
>    - *M* sql/lease.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-28>
>    (61)
>    - *M* sql/plan.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-29>
>    (87)
>    - *M* sql/rename.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-30>
>    (147)
>    - *M* sql/rowfetcher.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-31> (5)
>    - *M* sql/rowwriter.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-32>
>    (12)
>    - *M* sql/schema_changer.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-33>
>    (68)
>    - *M* sql/session.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-34>
>    (19)
>    - *M* sql/set.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-35>
>    (18)
>    - *M* sql/show.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-36>
>    (70)
>    - *M* sql/sqlutil/internal_executor.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-37> (7)
>    - *M* sql/table.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-38>
>    (69)
>    - *M* sql/truncate.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-39>
>    (10)
>    - *M* sql/update.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-40> (9)
>    - *M* storage/id_alloc.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-41>
>    (14)
>    - *M* storage/intent_resolver.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-42>
>    (25)
>    - *M* storage/log.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-43>
>    (15)
>    - *M* storage/raft_log_queue.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-44> (2)
>    - *M* storage/range_tree.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-45>
>    (457)
>    - *M* storage/replica_command.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-46>
>    (34)
>    - *M* storage/replica_gc_queue.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-47> (6)
>    - *M* storage/store.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-48>
>    (11)
>    - *M* ts/db.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-49> (2)
>    - *M* ts/query.go
>    <https://github.com/cockroachdb/cockroach/pull/6367/files#diff-50>
>    (14)
>
> Patch Links:
>
>    - https://github.com/cockroachdb/cockroach/pull/6367.patch
>    - https://github.com/cockroachdb/cockroach/pull/6367.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6367>
>

andreimatei commented:
getting rid of `pErr` above the KV client so we don't have to create `ErrorDetails` every time we want to have an error type.

---

Review status: 0 of 51 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGSPbyrRx3mYrp5sRUx:185367573)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
thanks

---

Review status: 0 of 51 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGSVPFyMsOJrFmajmi4:-1221266418)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Looks good so far. I've tried to comment on all the places which you've highlighted as sketchy and scrolled over the innocuous-looking swaths of diff. Let me know if there's any part that you think I've missed or brushed over.

One question that I have is whether you're planning to keep that `RetryableError`-sensitive code in `NewError`. That looks like an antipattern. With that gone (and some of your TODOs addressed, for instance removing "transaction" from RetryableTxnError - it should only hold the TxnID to allow for assertions), I think we're in a good spot.

---

Reviewed 3 of 3 files at r1, 38 of 38 files at r2, 22 of 22 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 575 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSrihnu5NSplO685OJ:-KGSrihnu5NSplO685OK:993048844) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/client/txn.go#L575)):*
you're doing the same assertion before, consider deduping (ignore if done in later rev)

---

*[kv/txn_test.go, line 425 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSvzfeR4fbD8jlWyUE:-KGSvzffUR_WBRiwA4T2:1734532739) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/kv/txn_test.go#L425)):*
if you can't type assert here, how do you type assert a `ConditionalPutError`? What's the type of `err`?

---

*[roachpb/errors.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSrupuF7VnPw6V-F7K:-KGSrupuF7VnPw6V-F7L:1795452209) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/roachpb/errors.go#L28)):*
This might be better as an interface (with an internal implementation). The upshot is that if we start wrapping errors, we can make sure the interface remains implemented appropriately. Don't need to change it now - TODO is enough for me.

---

*[roachpb/errors.go, line 92 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwJDjCSDuK-BiIgjn:-KGSwJDjCSDuK-BiIgjo:672012984) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/roachpb/errors.go#L92)):*
Having to construct `*Error` from a `RetryableTxnError` indicates that something is going wrong. Where is this needed?

---

*[sql/lease.go, line 624 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSssjLQ6LV-vp11lTy:-KGSssjLQ6LV-vp11lTz:1248621486) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/sql/lease.go#L624)):*
can inline

---

*[sql/plan.go, line 203 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwheOQFOK3EooTOl5:-KGSwheOQFOK3EooTOl6:-793666107) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/sql/plan.go#L203)):*
Why? WIP?

---

*[storage/intent_resolver.go, line 219 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGStOHBhpgGH4UgWVXd:-KGStOHCsEl41STU7JI5:-156895708) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/intent_resolver.go#L219)):*
I think the right solution here is not to use `InternalAddRequest` but to send the request manually (requires access to `TxnCoordSender`, so plumbing, etc but who cares.
I *think* this is ok for the time being though - the push pErr does not contain an updated pusher, but only the pushee (and that remains in `err`).
Definitely warrants a fat TODO for cleanup, but I have some hope that this will continue working.

---

*[storage/store.go, line 1698 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSuo8-djpThIxPBTIg:-KGSuo8-djpThIxPBTIh:175918974) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/store.go#L1698)):*
Again it's janky because of the loss of `pErr` within `processWriteIntentError`. I think the goal should be to make `processWriteIntentError` return `*Error`. I have hopes of this being correct (mostly because on WriteIntentError we update the txn below anyway), but on a write/write conflict (TransactionPushError) *maybe* we lose *something*. If tests seem stable I'm comfortable with a big TODO (same as before, actually).

---

*[storage/store.go, line 1766 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSvB4mJZiHbBHMEwSw:-KGSvB4mJZiHbBHMEwSx:-250373407) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/store.go#L1766)):*
We shouldn't catch an error here (unless it's from the abort cache, in which case we would not get the crucial information that we've been aborted; instead we'll go around thinking we're still PENDING, potentially caught in an infinite loop).
Same issue: we must not used `RunWithResponse` on this level - we're trying to do internal kv stuff through the public interface.
Likely not exercised in tests, so I'd be ok tackling this separately.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGSxBSUUxIVncWJhucg:1253215791)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 1766 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSvB4mJZiHbBHMEwSw:-KGT-qfkeCCiuRS0_Wen:1926442076) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/store.go#L1766)):*
scenario:

- we're aborted and don't know it
- we have a read-write conflict, the push above fails and we get a WriteIntentError
- we try to update our transaction (right here, and if we don't we might be stuck in a race, that's why we do this - the txn proto we're using might be outdated)
- query fails because our home range has the abort cache populated
- we catch a transactionabortederror, but with a pending transaction (since we lose the original txn, and you just use the txn we had...)

so something is sketchy here, but it should all resolve nicely when we don't use `s.db` for these internal requests any more.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 575 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSrihnu5NSplO685OJ:-KGTHtPHq16wVljG3fnO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/client/txn.go#L575)):*
Done.

---

*[kv/txn_test.go, line 425 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSvzfeR4fbD8jlWyUE:-KGTLoLdwQtDanMgcmX0:2029515546) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/kv/txn_test.go#L425)):*
you're right, we can type assert

---

*[roachpb/errors.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSrupuF7VnPw6V-F7K:-KGTDFs9e6btoKcCD5b9:324345191) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/roachpb/errors.go#L28)):*
Not sure I follow... We're not wrapping any errors, at least so far. So let's make it an interface if/when we need to.

---

*[roachpb/errors.go, line 92 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwJDjCSDuK-BiIgjn:-KGTEWyk8XUZBYrkJyKG:-661484336) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/roachpb/errors.go#L92)):*
you've seen where; added a TODO

---

*[sql/lease.go, line 624 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSssjLQ6LV-vp11lTy:-KGTEl03GPlnrRwKiStn:-661307734) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/sql/lease.go#L624)):*
let's stop inlining for blocks this big; it's not readable

---

*[sql/plan.go, line 203 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwheOQFOK3EooTOl5:-KGTEv5u8IdoyiHQheO4:-1901653472) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/sql/plan.go#L203)):*
yes, WIP, but I'll leave it as such for this PR. I still had to cut short the pErr - err transition in a couple of places to save this PR from getting too humongous and keeping `makePlan` return `pErr` is such a place.

---

*[storage/intent_resolver.go, line 219 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGStOHBhpgGH4UgWVXd:-KGTGFNLE4GFLQt3MAzl:1136123690) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/intent_resolver.go#L219)):*
added TODO. I still don't understand Kenji's TODO above but I hope you do.

---

*[storage/store.go, line 1698 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSuo8-djpThIxPBTIg:-KGTGcYpvnZzNSpTZI3B:2018543432) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/store.go#L1698)):*
added a pair TODO

---

*[storage/store.go, line 1766 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSvB4mJZiHbBHMEwSw:-KGTHFNJEsjUVPQLFNvV:284774471) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/store.go#L1766)):*
pasted all this in a TODO

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 46 of 51 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSrupuF7VnPw6V-F7K:-KGTRdqLZX8HThHUm3zc:772096618) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/roachpb/errors.go#L28)):*
OK, it turns out that it'd be good to wrap errors after all. There are KV tests that do things like txn.Put() and check that they got a `TransactionPushError` back.
So I'll indeed turn this into an interface which is going to be implemented by a couple of `ErrorDetail`s. And I'll have `Error.GoError()` return the error detail, if one is present.
SGTY? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 562 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGTWHOavIUdA93VAlsW:-KGTWHOavIUdA93VAlsX:1202076973) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6a3451dfb2a2338b46a12f2af2403e01099b6aa/client/txn.go#L562)):*
nitnitnit: checking the assertion below doesn't hurt even when we have no intention of retrying.

---

*[roachpb/errors.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSrupuF7VnPw6V-F7K:-KGTVNb7bUjx65Cc_hls:-542041061) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/roachpb/errors.go#L28)):*
Yes, that sounds good to me.

---

*[roachpb/errors.go, line 92 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwJDjCSDuK-BiIgjn:-KGTWT1htnCq7sA7-TQR:-1664238458) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/roachpb/errors.go#L92)):*
For my peace of mind, could this bit of code live in a clearly marked temp function that takes the output of `roachpb.NewError` and then performs surgery? That way, we know it's not active anywhere unintentionally, plus a good home for your TODO here.

---

*[sql/lease.go, line 624 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSssjLQ6LV-vp11lTy:-KGTVdS1kdbkolczYu7e:-527345603) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/sql/lease.go#L624)):*
I disagree, but it's obviously a nit you can ignore.

---

*[sql/plan.go, line 203 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwheOQFOK3EooTOl5:-KGTVkzSMIpB5jMDV2Ix:-228677403) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/sql/plan.go#L203)):*
Fine by me as long as there aren't any dragons here.

---

*[storage/intent_resolver.go, line 219 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGStOHBhpgGH4UgWVXd:-KGTVvM-5OImP1yUob6N:1308604866) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/intent_resolver.go#L219)):*
You can remove Kenji's TODO. nit: I use `tschottdorf` for my TODOs (and I sometimes grep for them).

---

*[storage/intent_resolver.go, line 221 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGTWxzqtsN2Ca53t1MT:-KGTWxzqtsN2Ca53t1MU:2054937052) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6a3451dfb2a2338b46a12f2af2403e01099b6aa/storage/intent_resolver.go#L221)):*
s/Dist/TxnCoord/

---

*[storage/store.go, line 1766 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSvB4mJZiHbBHMEwSw:-KGTW5_AR4bbsHWrjDY8:1144061168) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/store.go#L1766)):*
:thumbsup: TODO(tschottdorf) please though :)

---

*[storage/store.go, line 1773 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGTX2iT2zWPJ5wWurM0:-KGTX2iT2zWPJ5wWurM1:-1999719422) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6a3451dfb2a2338b46a12f2af2403e01099b6aa/storage/store.go#L1773)):*
nitnit: double space

---

*[storage/store.go, line 1779 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGTX61YDPnSsBw7o0EZ:-KGTX61YDPnSsBw7o0E_:-1274352268) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6a3451dfb2a2338b46a12f2af2403e01099b6aa/storage/store.go#L1779)):*
nit: can't tell from reviewable, but may want to reflow.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I've gotten the tests in `kv` and `client` to pass.
PTAL at `roachpb/error.go`.

I've deleted one test that I didn't really know what to do about (TestBeginTransactionErrorIndex) and I've opened an issue to rewrite another (`TestNestedTransaction`) that I've also removed (but this one maybe temporarily).

---

Review status: 43 of 55 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 92 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwJDjCSDuK-BiIgjn:-KGUIIrLJzy6t5chW7OG:204084792) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/roachpb/errors.go#L92)):*
umm I don't think so... The output of `NewError` has lost track of whether we had a retryable error... 
I think if you want peace of mind, you'll do your TODOs and I'll do mine and then this monstrosity can go away. But I'm open to suggestions.

---

*[storage/intent_resolver.go, line 219 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGStOHBhpgGH4UgWVXd:-KGUIsPd69Vj2ZY7VPN5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/intent_resolver.go#L219)):*
Done.

---

*[storage/intent_resolver.go, line 221 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGTWxzqtsN2Ca53t1MT:-KGUIvz5RopjyV2Ce2_a:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6a3451dfb2a2338b46a12f2af2403e01099b6aa/storage/intent_resolver.go#L221)):*
Done.

---

*[storage/store.go, line 1766 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSvB4mJZiHbBHMEwSw:-KGUJ2Pk31e4CuPPOKyo:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb789a06997bde88bae7d0a0cba8be07d1b7868a/storage/store.go#L1766)):*
Done.

---

*[storage/store.go, line 1773 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGTX2iT2zWPJ5wWurM0:-KGUJG-8bxTgiBtvBmX2:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6a3451dfb2a2338b46a12f2af2403e01099b6aa/storage/store.go#L1773)):*
Done.

---

*[storage/store.go, line 1779 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGTX61YDPnSsBw7o0EZ:-KGUJIrl85YdtZviaY8g:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6a3451dfb2a2338b46a12f2af2403e01099b6aa/storage/store.go#L1779)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGUJtwEqLzpyX0iRGei:1114277486)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 43 of 55 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 92 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwJDjCSDuK-BiIgjn:-KGVQGJwbIO_z1po5NOW:66643036) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/roachpb/errors.go#L92)):*
You would call the helper with the output of NewError and its input.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
OK I've gotten all the tests to pass.
I had to add yet another hack - keeping track of the `index` that is part of pErr (index of the error within the batch results) and move it forward to one of the error details. That's because my previous attempt to stop using the index and infer what command errored based on the first result that had an error didn't work for reasons that I don't really understand. There's code that backfills an error to every single result of a batch, for some reason, and trying to get rid of it makes some assertions fire. So I'll have to elucidate offline.

---

Review status: 39 of 76 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 92 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwJDjCSDuK-BiIgjn:-KGeY3qbseGAKocP2wW9:-1820011440) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/roachpb/errors.go#L92)):*
yeah... truth is there's still a bunch of places where this surgery is needed (a bunch of places in the descriptor leasing code, for example), and I'm afraid I'd miss some if I make it a separate function. Let's just aim to get rid of it quick.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGeYPrC8W3SRCkRV3Z3:-1112178200)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
So close!

---

Reviewed 12 of 12 files at r6, 28 of 28 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[client/db_test.go, line 420 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfV4crqynRZia3F2U9:-KGfV4crqynRZia3F2UA:1418875026) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbf0b625f0969d548d033016e6f28273045b2d9/client/db_test.go#L420)):*
We still do have a test that exercises the assertion around the transaction ID, right?

---

*[client/txn_test.go, line 334 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfVFqJaHBvCcwzs-uP:-KGfVFqKfXWz2s8ah-lg:-2001405593) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbf0b625f0969d548d033016e6f28273045b2d9/client/txn_test.go#L334)):*
At least while you don't fully remove the index, can this test remain?

---

*[roachpb/errors.go, line 92 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwJDjCSDuK-BiIgjn:-KGfUccvi469RrSEwLs1:1649368245) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/roachpb/errors.go#L92)):*
So? Honestly makes me much less uncomfortable to have some hack-helper that you call in a bunch of places that you may be forgetting in a place or two, than having code that runs when you don't know about it.

---

*[roachpb/errors.go, line 29 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfVsN5QVxIJ8Hn5O7v:-KGfVsN5QVxIJ8Hn5O7w:1026142446) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcf4301351f674ef3acdfe484c00c3d24f982905/roachpb/errors.go#L29)):*
Why the pointer?

---

*[roachpb/errors.go, line 41 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfW6ohBQbM2FadXbIG:-KGfW6ohBQbM2FadXbIH:1705225216) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcf4301351f674ef3acdfe484c00c3d24f982905/roachpb/errors.go#L41)):*
Why have this *and* Cause?

---

*[roachpb/errors.proto, line 161 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfXf3nzJ_H0fL8cuWJ:-KGfXf3nzJ_H0fL8cuWK:1201236040) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c67d1134cecd8cfae8753c97d38462fee28d1a/roachpb/errors.proto#L161)):*
Hrm, this is a regression to some extent (we've had this field before and decided that it wasn't going to remain). I see the problem with having to associate the error with a specific slot in the batch (the introduction of the Index field we have now took a while and was primarily motivated by exactly this). I understand that you want to draw the line somewhere, but changing protos isn't something we should take too lightly, especially in the absence of a plan.

What if you let `(Result).PErr` alone (i.e. don't make it an error) and added `(Result).Err() error` for anyone who wants to consume an error out of a result? Then the batch conversion code could use the `PErr` and you wouldn't need a hack. That seems like a good middle ground between exposing all the information to code which sits at the kv-sql interface and not spewing *Error outside of kv's realm.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGfbE0YirR_kr2R3VtZ:879131624)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.proto, line 161 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfXf3nzJ_H0fL8cuWJ:-KGgjf6M6DLv6OSABmPL:1963379634) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c67d1134cecd8cfae8753c97d38462fee28d1a/roachpb/errors.proto#L161)):*
So what's the problem with associating the error with one slot?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.proto, line 161 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfXf3nzJ_H0fL8cuWJ:-KGhOqBM9sTw3XsrOe3R:1089012291) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c67d1134cecd8cfae8753c97d38462fee28d1a/roachpb/errors.proto#L161)):*
The Index is semantically higher-level and so it's very unintuitive when to handle it (certainly not when creating the error, but at some random later point in time). Plus, you'll have the same issue again should we ever care about any other type of error's position. Take a look at the `Result.PErr` idea above. If it works out as I think, it'll be very nice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
please take a look at the last commit, adding PErr back to batch results.
There's some sketchiness in `client/db.go` in `runOneResult / runOneRow`, see if you have any suggestion for that code 

---

Review status: 69 of 76 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 41 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfW6ohBQbM2FadXbIG:-KGmDDuTmWmUwNSMAOFz:1612323567) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcf4301351f674ef3acdfe484c00c3d24f982905/roachpb/errors.go#L41)):*
`Cause` is used by tests to assert what exactly caused a retryable error. I think that's here to stay.
`CauseProto` is used to go back to a pErr from an error (and so should go away). In particular, those pesky `TransactionPushError`s that need to go pErr -> err -> pErr -> err and the final err needs to be asserted by tests to be TransactionPushError. Without it I'm not sure how I'd go back to a pErr with the correct detail.

---

*[roachpb/errors.proto, line 161 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfXf3nzJ_H0fL8cuWJ:-KGmnEMEgfN57wZudXUJ:1457451839) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c67d1134cecd8cfae8753c97d38462fee28d1a/roachpb/errors.proto#L161)):*
ok gave it a try

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGmnJSR0XP1hGZKH1pF:1824420271)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Could you highlight the sketchiness? Didn't see anything outrageous, but want to make sure.

---

Reviewed 7 of 7 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 124 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGmoir7aYt9LtZ7bN9q:-KGmoir9fLAnSVDaMGMq:-792050252) ([raw file](https://github.com/cockroachdb/cockroach/blob/4be48ca187535ab5fe1c1b6c8b5e7cd138e52858/client/db.go#L124)):*
s/use the Err() function/use Err() instead/

---

*[client/db.go, line 502 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGmoufvYsogm8pYPW8b:-KGmoufvYsogm8pYPW8c:79594604) ([raw file](https://github.com/cockroachdb/cockroach/blob/4be48ca187535ab5fe1c1b6c8b5e7cd138e52858/client/db.go#L502)):*
Total lie about the parallelism, I think. Unless the rpc sender does funny things. Anyhoo, maybe add a TODO here to clean up the comment.

---

*[client/db.go, line 531 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGmpS1c2Xpd8wCONlL-:-KGmpS1c2Xpd8wCONlL0:566803768) ([raw file](https://github.com/cockroachdb/cockroach/blob/4be48ca187535ab5fe1c1b6c8b5e7cd138e52858/client/db.go#L531)):*
You could call `runOneResult` which would avoid duplicating the assertion and a hair of code above.

---

*[roachpb/errors.go, line 41 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfW6ohBQbM2FadXbIG:-KGmqyrht441y7rjovlz:-1442951207) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcf4301351f674ef3acdfe484c00c3d24f982905/roachpb/errors.go#L41)):*
but can't you put a CauseProto into a Cause already? I though they all implemented ErrorDetailInterface, so you really wouldn't need the temporary field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGmrAYf31jEzGQb4LdN:-1684865117)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 22 of 38 files at r2, 16 of 22 files at r3, 7 of 12 files at r6, 24 of 28 files at r7, 7 of 7 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 502 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGmoufvYsogm8pYPW8b:-KGqEdRIFz-hX2TD8Pyx:-1044368516) ([raw file](https://github.com/cockroachdb/cockroach/blob/4be48ca187535ab5fe1c1b6c8b5e7cd138e52858/client/db.go#L502)):*
"The operations within a batch are run in an unspecified order and may be run in parallel". The point of this comment was to warn the caller not to rely on serial execution because we intend to parallelize it in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
added a new commit that tries to do some crap around not propagating retryable errors for the wrong transactions. It's pretty best effort; please see what you think

---

Review status: 67 of 76 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[client/db_test.go, line 420 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfV4crqynRZia3F2U9:-KGsSIg_EoZbjLBmQlsS:473058273) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbf0b625f0969d548d033016e6f28273045b2d9/client/db_test.go#L420)):*
rewrote this test

---

*[roachpb/errors.go, line 29 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfVsN5QVxIJ8Hn5O7v:-KGsSLX0Hur9Dk0p5lKa:-230258387) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcf4301351f674ef3acdfe484c00c3d24f982905/roachpb/errors.go#L29)):*
cause it's optional

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGsVWowboMG8pqlZ4U2:51559530)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, how about that rebase? I think this not-propagating-rabbithole is deep enough now.

---

Reviewed 9 of 9 files at r9.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 502 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGmoufvYsogm8pYPW8b:-KGqMwh6yiFElxDfIKPS:-270291300) ([raw file](https://github.com/cockroachdb/cockroach/blob/4be48ca187535ab5fe1c1b6c8b5e7cd138e52858/client/db.go#L502)):*
With the API private, seems better to remove this or at least soften it down. I doubt we've been paying attention to this in existing code, so we'd need a full audit anyway.

---

*[client/db.go, line 459 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGus4Fo6YiWqeoRSbE4:-KGus4FrLUPnDScQiq40:1356456428) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/db.go#L459)):*
How can one get here when youre `Exec`ing with `AutoRetry: true`?

---

*[client/db_test.go, line 437 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGusXkmw20W9vkUU7Q0:-KGusXkmw20W9vkUU7Q1:-523339212) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/db_test.go#L437)):*
Don't you want to assert that you don't get the RetryableTxnError? Matching the string is probably ok, but even with the RetryableTxnError I could imagine seeing the Cause in the string output.

---

*[client/txn.go, line 520 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGuso5nJ3wWI6YWBpKp:-KGuso5nJ3wWI6YWBpKq:-651404806) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/txn.go#L520)):*
s/ius/only/

---

*[client/txn.go, line 521 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGusrVVrxv8vnrtVRx5:-KGusrVVrxv8vnrtVRx6:-1706250519) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/txn.go#L521)):*
s/tt/t/

---

*[client/txn.go, line 526 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGut1usE7xlvCIWaV5C:-KGut1usE7xlvCIWaV5D:-206734363) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/txn.go#L526)):*
Why not stringify the error?

---

*[roachpb/errors.go, line 272 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGutRLAzWWnwj4j_dVa:-KGutRLAzWWnwj4j_dVb:676408629) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/roachpb/errors.go#L272)):*
? :)

---

*[roachpb/errors.proto, line 158 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGutSieId4xZR4nZ_KR:-KGutSieId4xZR4nZ_KS:677940215) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/roachpb/errors.proto#L158)):*
ping

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-:-KGutZFaJlRT1Zt9byN4:-736430676)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 78 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[client/db.go, line 124 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGmoir7aYt9LtZ7bN9q:-KGxE622PqwADq82CKPI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4be48ca187535ab5fe1c1b6c8b5e7cd138e52858/client/db.go#L124)):*
Done.

---

*[client/db.go, line 502 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGmoufvYsogm8pYPW8b:-KGwcvdDsDas_oF1GQ6W:2040859699) ([raw file](https://github.com/cockroachdb/cockroach/blob/4be48ca187535ab5fe1c1b6c8b5e7cd138e52858/client/db.go#L502)):*
I'm not sure what best to do to the comment. Please change it to your heart's desire (github now lets you edit files on the web!).

---

*[client/db.go, line 531 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGmpS1c2Xpd8wCONlL-:-KGwdto19RtevtnKaoqC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4be48ca187535ab5fe1c1b6c8b5e7cd138e52858/client/db.go#L531)):*
Done.

---

*[client/db.go, line 459 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGus4Fo6YiWqeoRSbE4:-KGwP0c6Cym_QFIDgUeD:-1821578760) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/db.go#L459)):*
yeah I've debated adding this here or not. You can still get the error here if we've run out of retries according to the default retry policy. Which might not be a thing right now because I think I think we don't set a retry count limit, but we might add a timeout or something in the future. So I dunno, it seemed like a good idea. At the very least, the code in Exec() is written as if the retry loop could stop at some point. 

---

*[client/db_test.go, line 437 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGusXkmw20W9vkUU7Q0:-KGwdzixDUznIl2BUy3y:-1002052865) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/db_test.go#L437)):*
Indeed the string doesn't change between RetryableError and another error.
I don't know if we should assert we're not getting RetryableError specifically. The more important assertion is the one below, that we don't retry. I'd just leave it like this.

---

*[client/txn.go, line 520 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGuso5nJ3wWI6YWBpKp:-KGweNck3hMNiQjVSLiD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/txn.go#L520)):*
Done.

---

*[client/txn.go, line 521 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGusrVVrxv8vnrtVRx5:-KGweOUcTRMIQX4N34Dc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/txn.go#L521)):*
Done.

---

*[client/txn.go, line 526 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGut1usE7xlvCIWaV5C:-KGwOHSmtQ3aostftb8a:-12914517) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/client/txn.go#L526)):*
well we can't strigify because sql handles `RetryableTxnError`. So sql is responsible for not propagating the error to someone else who might use it in the wrong context. So this here is just (some) protection against `txn.Exec()` getting the same error again.
`DB.Txn()` stringifies though. 

---

*[client/txn_test.go, line 334 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfVFqJaHBvCcwzs-uP:-KGwggCzIwfWZlzhCKQG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbf0b625f0969d548d033016e6f28273045b2d9/client/txn_test.go#L334)):*
Done.

---

*[roachpb/errors.go, line 92 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGSwJDjCSDuK-BiIgjn:-KGwiGFkTavevqyR49mT:827001251) ([raw file](https://github.com/cockroachdb/cockroach/blob/a080423b054b816ee51c1c6969da12084d56463a/roachpb/errors.go#L92)):*
I've added a TODO to move it to its own function (besides the TODO to delete it all-together :) ), once sql breaks free from pErr. Before then I truly don't know where to add the magic call.

---

*[roachpb/errors.go, line 41 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGfW6ohBQbM2FadXbIG:-KGwi9yz9Jl5C1VR-qJB:160413355) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcf4301351f674ef3acdfe484c00c3d24f982905/roachpb/errors.go#L41)):*
meh

---

*[roachpb/errors.go, line 272 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGutRLAzWWnwj4j_dVa:-KGxEw85PTk3O2uzcMrn:-1054325829) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/roachpb/errors.go#L272)):*
removed

---

*[roachpb/errors.proto, line 158 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367#-KGutSieId4xZR4nZ_KR:-KGxEoy6J0Rup-6Is-1j:1551689996) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6c9b8c08492ae52a1d5601bb6cfdf8e44227274/roachpb/errors.proto#L158)):*
removed all the `!!!`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6367)*
<!-- Sent from Reviewable.io -->
